File: q126-150/q128.py
# 2020.02.15
# Problem Statement:
# https://leetcode.com/problems/longest-consecutive-sequence/
# I was stupid enough to fail to come up with O(n) solution :-(
# https://leetcode.com/problems/longest-consecutive-sequence/discuss/41057/Simple-O(n)-with-Explanation-Just-walk-each-streak
# Really wanna quote this comment under the discussion page above:
# "I really hate you. Every time your solution makes mine look like shit. EVERY TIME."

class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        numsSet = set(nums)
        answer = 0
        for element in numsSet:
            # if this element is the start or sequence
            if element - 1 not in numsSet:
                next = element + 1
                # check if the next still exists
                while next in numsSet:
                    next = next + 1
                # compare with the current best solution
                answer = max(answer, next-element)
                
        return answer

File: Blind75/q128.java
// 2022.09.16
// Problem Statement:
// https://leetcode.com/problems/longest-consecutive-sequence/

// idea: dfs with cache
class Solution {
    public HashMap<Integer, Integer> ht;
    public int answer = 0;
    public int dfs(int n) {
        if (!ht.containsKey(n)) return 0;
        if (ht.containsKey(n) && ht.get(n)>0) {
            return ht.get(n);
        }
        
        int res = 1+dfs(n+1);
        ht.put(n, res);
        answer = Math.max(answer, res);
        return res;
    }
    
    public int longestConsecutive(int[] nums) {
        ht = new HashMap<> ();
        for (int i=0; i<nums.length; i++) {
            ht.put(nums[i], 0);
        }
        
        for (int i=0; i<nums.length; i++) {
            int x = dfs(nums[i]);
        }
        return answer;
    }
}